subsys/net/lib/lwm2m/lwm2m_object.h

```diff
@@ -132,8 +132,18 @@
 BUILD_ASSERT(CONFIG_LWM2M_COAP_BLOCK_SIZE <= CONFIG_LWM2M_COAP_MAX_MSG_SIZE,
 	     "CoAP block size can't exceed maximum message size");
 
-#define MAX_PACKET_SIZE		(CONFIG_LWM2M_COAP_MAX_MSG_SIZE + \
-				 CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE)
+BUILD_ASSERT(CONFIG_LWM2M_COAP_MAX_MSG_SIZE % CONFIG_LWM2M_COAP_BLOCK_SIZE == 0,
+	     "CoAP maximum message size needs to be multiple of the block size");
+
+#define LWM2M_COAP_BLOCK_TRANSFER_ENABLED                                                          \
+	(CONFIG_LWM2M_COAP_BLOCK_SIZE < CONFIG_LWM2M_COAP_MAX_MSG_SIZE)
```

Comment:
I think this should simply be a Kconfig option. Plus probably another one to determine the actual size of the buffer for large message serialization.

Reply:
I'm happy to add a Kconfig for `LWM2M_COAP_BLOCK_TRANSFER_ENABLED`. Maybe anyone has a better name for the config.

But I'm a bit confused about the config for the big buffer. Then we would have tree different sizes for buffers:
1. block size
2. message size
3. big buffer size for serialization

Probably I don't understand the meaning of the message size.

Reply:
Message size represents the maximum CoAP payload size the message can carry. In most cases, it should be lower than network MTU (minus CoAP/UDP/IP header sizes) so that the message can fit a single UDP datagram. In case IP fragmentation is supported though, one could try to use even larger message sizes, however I don't think we've ever tested such case.

So to sum up:
* `LWM2M_COAP_BLOCK_SIZE` represents the CoAP block size size used by LwM2M, this is pretty straightforward
* `LWM2M_COAP_MAX_MSG_SIZE` represents the maximum payload a **single** CoAP packet can carry. This should be no lower than `LWM2M_COAP_BLOCK_SIZE`, and block transfer should only be involved if the desired payload size to transmit would exceed `LWM2M_COAP_MAX_MSG_SIZE`
* Plus, as you've noticed, we'd need a third config for the serialization buffer. This should be higher than `LWM2M_COAP_MAX_MSG_SIZE`, as otherwise it'd make little sense (whole serialization buffer would fit a single CoAP message, so there's no really need for block transfer).